### PR TITLE
bounded lattices in BloqadeLattices

### PR DESCRIPTION
**Pull request by weinbe58**
adding Bounded lattices to BloqadeLattices.

Our implementation involves a single new type `BoundedLattice` that combines an `AbstractLattice` and  an `AbstractRegion` 

PBC is enforced on the level of a `distance` function calculated by the `AbstractRegion`

### CODE REVIEW

File: lib/BloqadeLattices/src/lattice.jl [hunk 1]
```diff
@@ -51,7 +51,7 @@ lattice_vectors(general_lattice::GeneralLattice) = general_lattice.vectors
 """
     lattice_sites(lattice::AbstractLattice)
 
-Returns sites in a Bravais lattice unit cell as a Tuple of D-Tuple, where D is the space dimension.
+Returns sites in a Bravais lattice unit region as a Tuple of D-Tuple, where D is the space dimension.
```
Comment (weinbe58) [inline lib/BloqadeLattices/src/lattice.jl:54]:
```suggestion
Returns sites in a Bravais lattice unit cell as a Tuple of D-Tuple, where D is the space dimension.
```

File: lib/BloqadeLattices/src/lattice.jl [hunk 2]
```diff
@@ -30,10 +30,10 @@ end
     GeneralLattice(vectors, sites)
 
 The general lattice type for tiling the space. Type parameter `D` is the dimension,
-`K` is the number of sites in a unit cell and `T` is the data type for coordinates, e.g. `Float64`. Input arguments are
+`K` is the number of sites in a unit region and `T` is the data type for coordinates, e.g. `Float64`. Input arguments are
 
 * `vectors` is a vector/tuple of D-tuple. Its length is D, it specifies the Bravais lattice vectors.
-* `sites` is a vector/tuple of D-tuple. Its length is K, it specifies the sites inside a Bravais cell.
+* `sites` is a vector/tuple of D-tuple. Its length is K, it specifies the sites inside a Bravais region.
```
Comment (weinbe58) [inline lib/BloqadeLattices/src/lattice.jl:36]:
```suggestion
* `sites` is a vector/tuple of D-tuple. Its length is K, it specifies the sites inside a Bravais cell.
```

File: lib/BloqadeLattices/src/lattice.jl [hunk 3]
```diff
@@ -30,10 +30,10 @@ end
     GeneralLattice(vectors, sites)
 
 The general lattice type for tiling the space. Type parameter `D` is the dimension,
-`K` is the number of sites in a unit cell and `T` is the data type for coordinates, e.g. `Float64`. Input arguments are
+`K` is the number of sites in a unit region and `T` is the data type for coordinates, e.g. `Float64`. Input arguments are
```
Comment (weinbe58) [inline lib/BloqadeLattices/src/lattice.jl:33]:
```suggestion
`K` is the number of sites in a unit cell and `T` is the data type for coordinates, e.g. `Float64`. Input arguments are
```

File: lib/BloqadeLattices/src/BloqadeLattices.jl [hunk 4]
```diff
@@ -7,6 +7,7 @@ using StatsBase
 using LuxorGraphPlot
 using LuxorGraphPlot: Point
 using LuxorGraphPlot.Luxor: Colors
+using Combinatorics: combinations
```
Comment (weinbe58) [inline lib/BloqadeLattices/src/BloqadeLattices.jl:10]:
```suggestion
```
remove this dependency. It shouldn't be needed.

File: lib/BloqadeLattices/Project.toml [hunk 5]
```diff
@@ -3,6 +3,7 @@ uuid = "bd27d05e-4ce1-5e79-84dd-c5d7d508bbe4"
 version = "0.1.3"
 
 [deps]
+Combinatorics = "861a8166-3701-5b0c-9a16-15d98fcdc6aa"
```
Comment (weinbe58) [inline lib/BloqadeLattices/Project.toml:6]:
```suggestion
```

remove.

File: lib/BloqadeLattices/src/lattice.jl (hunk 3 above)
Comment (weinbe58) [reply lib/BloqadeLattices/src/lattice.jl:33]: did a replace all and accidentally changed some wording.

**Review by weinbe58 — COMMENTED** [review]
Doing some minor fixes here.

**Comment by johnzl-777** [comment]
There are two issues that probably need some more wrangling with:

1. \in for regions seems to include points that are definitely not part of the region (see the `in_region_sanity_check` test) 
2. `generate_sites_in_region` may fail to generate all valid lattice points bounded by a region because of how it initializes its stack. It's possible for a user defined region to not include all points within the unit cell touching the origin. While the function does account for all lattice vectors in generating the neighbor of a given point it might need some additional "help" for more complex lattices (like the Honeycomb or Kagome instances where you can have a site not conveniently touching a vector/made by just adding unit vectors).

A test has been developed to verify if problem 2 really does pop up but is currently blocked due to the recent issue with \in.

File: lib/BloqadeLattices/test/runtests.jl [hunk 6]
```diff
@@ -52,6 +52,94 @@ end
...
+end
+
+@testset "in_region_sanity_check" begin
+    rot_mat(θ) = [cos(θ) -sin(θ); sin(θ) cos(θ)]
+
+    a1 = [1.0, 0.0]
+    a2 = [0.5, 0.5 * sqrt(3)]
+    bounds = zeros((2,2))
+    bounds[1,:] .= rot_mat(deg2rad(50)) * a1 * 2
+    bounds[2,:] .= rot_mat(deg2rad(50)) * a2 * 2
```
Comment (weinbe58) [inline lib/BloqadeLattices/test/runtests.jl:137]: > 1. \in for regions seems to include points that are definitely not part of the region (see the `in_region_sanity_check` test) and not due to floating point error either

Ah I found the issue, you put the vectors into the rows of the matrix not the rows
```suggestion
    bounds[:,1] .= rot_mat(deg2rad(50)) * a1 * 2
    bounds[:,2] .= rot_mat(deg2rad(50)) * a2 * 2
```

**Review by weinbe58 — COMMENTED** [review]
> 2. `generate_sites_in_region` may fail to generate all valid lattice points bounded by a region because of how it initializes its stack. It's possible for a user defined region to not include all points within the unit cell touching the origin. While the function does account for all lattice vectors in generating the neighbor of a given point it might need some additional "help" for more complex lattices (like the Honeycomb or Kagome instances where you can have a site not conveniently touching a vector/made by just adding unit vectors).

I think this is fine because in principle the region can be arbitrarily as small. We should just go ahead with our discussion today of using the integer tuple for the visited list.

File: lib/BloqadeLattices/src/interact.jl [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    return two_body_interaction_matrix(atoms,(x,y)->C/distance(x,y)^6)
+end
+
+
+### implementation for bounded lattices
+
+
+function rydberg_interaction_matrix(lat::BoundedLattice{L,R},C::Real) where {L,R}
+    if lat.PBC
+        return two_body_interaction_matrix(lat.site_positions,(x,y)->C/distance(lat.region,x,y)^6)
```
Comment (weinbe58) [inline lib/BloqadeLattices/src/interact.jl:28]:
```suggestion
        return two_body_interaction_matrix(lat.site_positions) do x,y
            return C/distance(lat.region,x,y)^6
        end
```

File: lib/BloqadeLattices/src/interact.jl [hunk 8]
```diff
@@ -0,0 +1,34 @@
+
+
+
+
+function two_body_interaction_matrix(atoms,f)
```
Comment (weinbe58) [inline lib/BloqadeLattices/src/interact.jl:5]:
```suggestion
function two_body_interaction_matrix(f,atoms)
```

File: lib/BloqadeLattices/src/bounded_lattice.jl [hunk 9]
```diff
@@ -26,7 +26,7 @@ end
 
 
 
-function parallelepiped_region(lattice::AbstractLattice{D},M::Vararg{NTuple{D,Int},D};PBC::Bool=false) where D
+function parallelepiped_region(lattice::AbstractLattice{D},M::Vararg{NTuple{D,Real},D};PBC::Bool=false) where D
```
Comment (johnzl-777) [inline lib/BloqadeLattices/src/bounded_lattice.jl:29]: The vectors defining the bounded region can be tuples of Integers OR Floats, so I felt it suitable to just define the type as `Real`

**Pull request merged**